iris_ugrid/ugrid_cf_reader.py

```diff
@@ -0,0 +1,240 @@
+
+        # Generate list of excluded variable names.
+        exclude_vars = list(meshes.keys())
+
+        for mesh in meshes.values():
+            mesh_var = dataset.variables[mesh.mesh_name]
+            for attr in mesh_var.ncattrs():
+                if attr in _UGRID_LINK_PROPERTIES:
+                    exclude_vars.extend(mesh_var.getncattr(attr).split())
+
+        # Identify possible mesh dimensions and make a map of them.
+        meshdims_map = {}  # Maps {dimension-name: (mesh, mesh-location)}
+        for mesh in meshes.values():
+            mesh_var = dataset.variables[mesh.mesh_name]
+            if mesh.faces is not None:
+                # Work out name of faces dimension and record it.
+                if "face_dimension" in mesh_var.ncattrs():
+                    faces_dim_name = mesh_var.getncattr("face_dimension")
+                else:
+                    # Assume default dimension ordering, and get the dim name
+                    # from dims of a non-optional connectivity variable.
+                    faces_varname = mesh_var.face_node_connectivity
+                    faces_var = dataset.variables[faces_varname]
+                    faces_dim_name = faces_var.dimensions[0]
+                meshdims_map[faces_dim_name] = (mesh, "face")
+            if mesh.edges is not None:
+                # Work out name of edges dimension and record it.
+                if "edge_dimension" in mesh_var.ncattrs():
+                    edges_dim_name = mesh_var.getncattr("edge_dimension")
+                else:
+                    # Assume default dimension ordering, and get the dim name
+                    # from dims of a non-optional connectivity variable.
+                    edges_varname = mesh_var.edge_node_connectivity
+                    edges_var = dataset.variables[edges_varname]
+                    edges_dim_name = edges_var.dimensions[0]
+                meshdims_map[edges_dim_name] = (mesh, "edge")
+            if mesh.nodes is not None:
+                # Work out name of nodes dimension and record it.
+                # Get it from a non-optional coordinate variable.
+                nodes_varname = mesh_var.node_coordinates.split()[0]
+                nodes_var = dataset.variables[nodes_varname]
+                nodes_dim_name = nodes_var.dimensions[0]
+                meshdims_map[nodes_dim_name] = (mesh, "node")
+        self.meshdims_map = meshdims_map
+
+        # Create a CFReader object which skips the UGRID-related variables.
+        kwargs["exclude_var_names"] = exclude_vars
+        self.cfreader = iris.fileformats.cf.CFReader(
+            self.dataset, *args, **kwargs
+        )
+
+    def complete_unstructured_cube(self, cube):
+        """
+        Cube post-processing hook called by :func:`iris.fileformats.netcdf.load_cubes`, to add the unstructured info
+        onto a cube newly created by the `self.cfreader`.
+
+        Add the ".ugrid" property to a cube loaded with the `self.cfreader`.
```

Comment:
As discussed offline with @trexfeathers  ..
Since looking into changes for cube summary (string representations) (cf #3914), I'm fairly convinced now that we should really be using a Cube *subclass*, allowing us to overload methods, rather than just attaching this info as a random extra property.
But I still think this will do for a start.